service/history/workflow/context.go

```diff
@@ -279,6 +279,12 @@ func (c *ContextImpl) LoadMutableState(ctx context.Context, shardContext shard.C
 		if err != nil {
 			return nil, err
 		}
+
+		// NOTE: we can't not assigned the result directly to c.MutableState like below
+		// c.MutableState, err = NewMutableStateFromDB(...)
+		// Otherwise c.MutableState itself will not be nil, but can potentially point to a nil *MutableStateImpl,
+		// and causing NPE (e.g. when calling c.Clear()) or other unexpected behavior.
```

Comment:
Could you add comment that c.MutableState is an interface, and NewMutableStateFromDB() returns a pointer, thus it would end up be an non nil interface pointing to a nil pointer.

Reply:
Also, I think if we change the func NewMutableStateFromDB() to return MutableState interface, we would avoid this confusing behavior. Shall we do that?

Reply:
hmm it's not trivial to just switch to return interface, some other implementations update the mutable state struct returned by it.  
There's also the golang "accept interface and return struct" convention, we mostly follow in the codebase.